src/write_buffer_manager.rs

```diff
@@ -0,0 +1,188 @@
+///! `WriteBufferManager` is for managing memory allocation for one or more
+///! MemTables.
+use std::ptr::NonNull;
+use std::sync::Arc;
+
+use crate::{ffi, Cache};
+
+pub(crate) struct WriteBufferManagerWrapper {
+    pub(crate) inner: NonNull<ffi::rocksdb_write_buffer_manager_t>,
+}
+
+// Just like the types in `db_options`, these are only safe because
+// the underlying types are thread-safe. This thread-safety is not actually
+// documented, but an evaluation of the code revealed that all member
+// variables are protected by mutexes or atomics.
+unsafe impl Send for WriteBufferManagerWrapper {}
+unsafe impl Sync for WriteBufferManagerWrapper {}
+
+impl Drop for WriteBufferManagerWrapper {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::rocksdb_write_buffer_manager_destroy(self.inner.as_ptr());
+        }
+    }
+}
+
+// Note that we _could_ clone the underlying `std::shared_ptr`, but just storing
+// it in an `Arc` and have the underlying calls (like `rocksdb_options_set_write_buffer_manager`)
+// copy-construct the underlying `std::shared_ptr` is much easier, and is the same way
+// we do it for types like `db_options::Cache`.
+#[derive(Clone)]
+pub struct WriteBufferManager(pub(crate) Arc<WriteBufferManagerWrapper>);
+
+impl WriteBufferManager {
+    /// _buffer_size = 0 indicates no limit.
+    pub fn new(buffer_size: usize) -> WriteBufferManager {
+        WriteBufferManager(Arc::new(WriteBufferManagerWrapper {
+            inner: NonNull::new(unsafe {
+                ffi::rocksdb_write_buffer_manager_create(buffer_size, false)
+            })
+            .unwrap(),
+        }))
+    }
+
+    pub fn new_with_allow_stall(buffer_size: usize, allow_stall: bool) -> WriteBufferManager {
+        WriteBufferManager(Arc::new(WriteBufferManagerWrapper {
+            inner: NonNull::new(unsafe {
+                ffi::rocksdb_write_buffer_manager_create(buffer_size, allow_stall)
+            })
+            .unwrap(),
+        }))
+    }
+
+    // Parameters:
+    // _buffer_size: _buffer_size = 0 indicates no limit.
+
+    // cache_: if `cache` is provided, we'll put dummy entries in the cache and
+    // cost the memory allocated to the cache. It can be used even if _buffer_size
+    // = 0.
```

Comment:
1. stray leading/trailing underscores
2. "if `cache` is provided" does not make sense, since `cache` is not optional
3. What does "cost the memory allocated to the cache" mean?